packages/playwright/src/fixture.ts

```diff
@@ -0,0 +1,296 @@
+import assert from "assert";
+import { Errors } from "./error";
+
+export interface FixtureStepStart {
+  id: string;
+  apiName: string;
+  params: Record<string, any>;
+  stackTrace: ParsedStackTrace;
+}
+
+export interface TestIdData {
+  id: number;
+  attempt: number;
+  source: {
+    title: string;
+    scope: string[];
+  };
+}
+
+interface FixtureStepStartEvent extends FixtureStepStart {
+  event: "step:start";
+  test: TestIdData;
+}
+
+export interface FixtureStepEnd {
+  id: string;
+  error: Error | undefined;
+}
+
+interface FixtureStepEndEvent extends FixtureStepEnd {
+  event: "step:end";
+  test: TestIdData;
+}
+
+interface ReporterErrorEvent extends ReporterError {
+  event: "error";
+  test: TestIdData;
+}
+
+export type FixtureEvent = FixtureStepStartEvent | FixtureStepEndEvent | ReporterErrorEvent;
+
+const debug = dbg("replay:playwright:fixture");
+
+function ReplayAddAnnotation([event, id, data]: any) {
+  // @ts-ignore
+  window.__RECORD_REPLAY_ANNOTATION_HOOK__?.("replay-playwright", {
+    event,
+    id,
+    data: data ? JSON.parse(data) : undefined,
+  });
+}
+
+function getLastStepId(testInfo: any) {
+  return testInfo._steps[testInfo._steps.length - 1].stepId;
+}
+
+function isReplayAnnotation(params?: any) {
+  return params?.expression?.includes("ReplayAddAnnotation");
+}
+
+function parseLocation(stack?: string) {
+  const pattern = /\/([^\/]+):(\d+):(\d+)$/;
+  const firstLine = stack?.split("\n").find(l => pattern.test(l));
+  const match = firstLine?.match(pattern);
+
+  if (!match) {
+    return {
+      line: undefined,
+      column: undefined,
+    };
+  }
+
+  return {
+    line: parseInt(match[2]),
+    column: parseInt(match[3]),
+  };
+}
+
+function parseError<T extends { name?: string; message: string; stack?: string }>(error: T) {
+  if (!error) {
+    return;
+  }
+
+  const location = parseLocation(error.stack);
+
+  return {
+    name: error.name || "Error",
+    message: error.message,
+    line: location?.line,
+    column: location?.column,
+  };
+}
+
+function maybeMonkeyPatchTestInfo(
+  testInfo: TestInfoInternal,
+  addStepHandler: (step: TestInfoStep) => void,
+  stepEndHandler: TestInfoInternal["_onStepEnd"]
+) {
+  if (testInfo._addStep) {
+    const original = testInfo._addStep.bind(testInfo);
+    testInfo._addStep = (data, parentStep) => {
+      const result = original(data, parentStep);
+      addStepHandler?.(result);
+
+      return result;
+    };
+  }
+
+  if (testInfo._onStepEnd) {
+    const original = testInfo._onStepEnd.bind(testInfo);
+    testInfo._onStepEnd = step => {
+      const result = original(step);
+      stepEndHandler?.(step);
+
+      return result;
+    };
+  }
+}
+
+export async function replayFixture(
+  { playwright, page }: { playwright: any; page: Page },
+  use: () => Promise<void>,
+  testInfo: TestInfo
+) {
+  debug("Setting up replay fixture");
+
+  const port = getServerPort();
+  const ws = new WebSocket(`ws://localhost:${port}`);
```

Comment:
Are we sure that the websocket server has already been started or do we need to eventually add some retry logic here?

Reply:
As far as I can tell, the reporter (where the server lives) always starts first.